guides/inventing-service.md

```diff
@@ -0,0 +1,919 @@
+headers.
+
+Since our `Handler` trait is useful for defining both servers and clients,
+`Handler` probably isn't an appropriate name anymore. A client doesn't handle an
+request, it sends it to a server and the server handles it. Lets instead call
+our trait `Service`:
+
+```rust
+trait Service<Request> {
+    type Response;
+    type Error;
+    type Future: Future<Output = Result<Self::Response, Self::Error>>;
+
+    fn call(&mut self, request: Request) -> Self::Future;
+}
+```
+
+This is actually _almost_ the `Service` trait defined in Tower. If you've been
+able to follow along until this point you now understand most of Tower. Besides
+the `Service` trait Tower also provides several utilities that implement
+`Service` by wrapping some other type that also implements `Service`. Exactly
+like we did with `Timeout` and `JsonContentType`. Things can be composed in ways
+similar to what we've done thus far.
+
+Some example services provided by Tower:
+
+- `Timeout` - This is pretty much identical to the timeout we have built.
+- `Retry` - To automatically retry failed requests.
+- `RateLimit` - To limit the number of requests a service can handle over a
+  period of time.
+
+Types like `Timeout` and `JsonContentType` are typically called "middleware"
+since they wrap another `Service` and transform the request or response in some
+way. Whereas types like `RequestHandler` is typically called a "leaf service"
+since it sits at the leaves of a tree of nested services. The actual responses
+are normally produced in leaf services and modified by middlewares.
+
+The only thing left to talk about is "backpressure" and `poll_ready`.
+
+## Backpressure
+
+Imagine you wanted to write a rate limit middleware that wraps a `Service` and
+puts a limit on the maximum number of concurrent requests the underlying service
+will receive. This would be useful if you had some service that had a hard upper
+limit on the amount of load it could handle.
+
+With our current `Service` trait we don't really have a good way to implement
+something like that. We could try:
+
+```rust
+impl<R, T> Service<R> for ConcurrencyLimit<T> {
+    fn call(&mut self, request: R) -> Self::Future {
+        // 1. Check a counter for the number of requests currently being
+        //    processed.
+        // 2. If there is capacity left send the request to `T`
+        //    and increment the counter.
+        // 3. If not somehow wait until capacity becomes available.
+        // 4. When the response has been produced, decrement the counter.
+    }
+}
+```
+
+If there is no capacity left we have to wait and somehow get notified when
+capacity becomes available. Additionally we have to keep the request in memory
+while we're waiting (also called buffering). It would be more robust to only
+allocate space for the request when we are sure the service has capacity to
+handle it. Otherwise we risk using a lot of memory buffering requests while we
+wait for our service to become ready.
```

Comment:
```suggestion
If there is no capacity left, we have to wait and somehow get notified when
capacity becomes available. Additionally, we have to keep the request in memory
while we're waiting (also called _buffering_). This means that the more requests
that are waiting for capacity, the more memory our program would use --- if more 
requests are produced than our service can handle, we might run out of memory! 
It would be more robust to only allocate space for the request when we are sure
the service has capacity to handle it. Otherwise, we risk using a lot of memory 
buffering requests while we wait for our service to become ready.
```